polaris-core/src/main/java/org/apache/polaris/core/persistence/transactional/TransactionalPersistence.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.persistence.transactional;
+
+import jakarta.annotation.Nonnull;
+import jakarta.annotation.Nullable;
+import java.util.List;
+import java.util.function.Supplier;
+import org.apache.polaris.core.PolarisCallContext;
+import org.apache.polaris.core.entity.EntityNameLookupRecord;
+import org.apache.polaris.core.entity.PolarisBaseEntity;
+import org.apache.polaris.core.entity.PolarisEntitiesActiveKey;
+import org.apache.polaris.core.entity.PolarisEntityCore;
+import org.apache.polaris.core.persistence.BasePersistence;
+import org.apache.polaris.core.persistence.IntegrationPersistence;
+
+/**
+ * Extends BasePersistence to express a more "transaction-oriented" control flow for backing stores
+ * which can support a runInTransaction semantic, while providing default implementations of some of
+ * the BasePersistence methods in terms of lower-level methods that subclasses must implement.
+ */
+public abstract class TransactionalPersistence implements BasePersistence, IntegrationPersistence {
```

Comment:
The common denominator for atomic operations in non-transactional databases is a CAS on a single row/key. Like "fetch all things", then "check all requirements", then "persist all the changes" and eventually "perform a single CAS" to perform the atomic change. How would this be implemented for example for a "rename table" operation or a "multi table commit" involving all the changes to the individual entities and consistency guarantees?